File: dataloom-frontend/src/api/projects.js
/**
 * API functions for project CRUD operations.
 * @module api/projects
 */
import client from "./client";

/**
 * Upload a new project dataset file.
 * @param {File} file - The dataset file to upload.
 * @param {string} projectName - Name for the new project.
 * @param {string} projectDescription - Description for the new project.
 * @returns {Promise<Object>} The created project response.
 */
export const uploadProject = async (file, projectName, projectDescription) => {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("projectName", projectName);
  formData.append("projectDescription", projectDescription);
  const response = await client.post("/projects/upload", formData);
  return response.data;
};

/**
 * Fetch full project details including rows and columns.
 * @param {string} projectId - The project ID.
 * @returns {Promise<Object>} Project details with columns and rows.
 */
export const getProjectDetails = async (projectId) => {
  const response = await client.get(`/projects/get/${projectId}`);
  return response.data;
};

/**
 * Fetch the most recently modified projects.
 * @returns {Promise<Array>} List of recent project summaries.
 */
export const getRecentProjects = async () => {
  const response = await client.get("/projects/recent");
  return response.data;
};

/**
 * Save the current project state as a checkpoint.
 * @param {string} projectId - The project ID.
 * @param {string} commitMessage - Description of changes.
 * @returns {Promise<Object>} Updated project response.
 */
export const saveProject = async (projectId, commitMessage) => {
  const response = await client.post(
    `/projects/${projectId}/save?commit_message=${encodeURIComponent(commitMessage)}`,
  );
  return response.data;
};

/**
 * Revert project to a previous checkpoint.
 * @param {string} projectId - The project ID.
 * @param {string} checkpointId - The checkpoint ID to revert to.
 * @returns {Promise<Object>} Reverted project response.
 */
export const revertToCheckpoint = async (projectId, checkpointId) => {
  const response = await client.post(`/projects/${projectId}/revert?checkpoint_id=${checkpointId}`);
  return response.data;
};

/**
 * Extract a filename from a Content-Disposition header if present.
 * @param {string | undefined} header
 * @returns {string | null}
 */
const getFilenameFromDisposition = (header) => {
  if (!header) return null;
  const match = header.match(/filename="?([^"]+)"?/i);
  return match ? match[1] : null;
};

/**
 * Export the current working copy of a project as a downloadable file.
 * @param {string} projectId - The project ID.
 * @param {string} format - Export format.
 * @returns {Promise<{blob: Blob, filename: string | null}>} The download payload.
 */
export const exportProject = async (projectId, format = "csv") => {
  const response = await client.get(`/projects/${projectId}/export`, {
    params: { format },
    responseType: "blob",
  });
  return {
    blob: response.data,
    filename: getFilenameFromDisposition(response.headers["content-disposition"]),
  };
};

/**
 * Download a generated quality report for a project.
 * @param {string} projectId - The project ID.
 * @param {string} format - Report format.
 * @returns {Promise<{blob: Blob, filename: string | null}>} The download payload.
 */
export const downloadQualityReport = async (projectId, format = "html") => {
  const response = await client.get(`/projects/${projectId}/quality-report`, {
    params: { format },
    responseType: "blob",
  });
  return {
    blob: response.data,
    filename: getFilenameFromDisposition(response.headers["content-disposition"]),
  };
};

/**
 * Delete a project and its associated files.
 * @param {string} projectId - The project ID.
 * @returns {Promise<Object>} Success confirmation.
 */
export const deleteProject = async (projectId) => {
  const response = await client.delete(`/projects/${projectId}`);
  return response.data;
};

File: dataloom-backend/app/api/endpoints/projects.py
"""Project CRUD API endpoints.

Handles upload, retrieval, save (checkpoint), and revert operations.
"""

import re
import uuid

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.responses import FileResponse, Response
from sqlmodel import Session

from app import database, models, schemas
from app.api.dependencies import get_project_or_404
from app.services.file_service import delete_project_files, get_original_path, store_upload
from app.services.project_service import (
    create_checkpoint,
    create_project,
    delete_project,
    get_recent_projects,
)
from app.services.quality_service import EXPORT_MEDIA_TYPES, REPORT_MEDIA_TYPES, build_quality_report
from app.services.transformation_service import apply_logged_transformation
from app.utils.logging import get_logger
from app.utils.pandas_helpers import (
    dataframe_to_bytes,
    dataframe_to_response,
    read_csv_safe,
    read_dataframe_safe,
    save_csv_safe,
)
from app.utils.security import validate_upload_file

logger = get_logger(__name__)

router = APIRouter()


@router.post("/upload", response_model=schemas.ProjectResponse)
async def upload_project(
    file: UploadFile = File(...),
    projectName: str = Form(...),
    projectDescription: str = Form(...),
    db: Session = Depends(database.get_db),
):
    """Upload a new dataset file for a project.

    Validates the file, stores it with a sanitized name, creates a working copy,
    and returns the initial project data.
    """
    logger.info("Upload request: project=%s, file=%s", projectName, file.filename)
    validate_upload_file(file)

    original_path, copy_path = store_upload(file)
    df = read_dataframe_safe(original_path)

    project = create_project(db, projectName, str(copy_path), projectDescription)

    resp = dataframe_to_response(df)
    return {
        "filename": project.name,
        "file_path": project.file_path,
        "project_id": project.project_id,
        **resp,
    }


@router.get("/get/{project_id}", response_model=schemas.ProjectResponse)
async def get_project_details(project_id: uuid.UUID, db: Session = Depends(database.get_db)):
    """Fetch full project details including all rows and columns."""
    project = get_project_or_404(project_id, db)
    df = read_csv_safe(project.file_path)

    resp = dataframe_to_response(df)
    return {
        "filename": project.name,
        "file_path": project.file_path,
        "project_id": project.project_id,
        **resp,
    }


@router.get("/recent", response_model=list[schemas.LastResponse])
def recent_projects(db: Session = Depends(database.get_db)):
    """Get the most recently modified projects."""
    projects = get_recent_projects(db, limit=10)
    return [
        schemas.LastResponse(
            project_id=p.project_id,
            name=p.name,
            description=p.description,
            last_modified=p.last_modified,
        )
        for p in projects
    ]


@router.post("/{project_id}/save", response_model=schemas.ProjectResponse)
async def save_project(
    project_id: uuid.UUID,
    commit_message: str,
    db: Session = Depends(database.get_db),
):
    """Save project changes as a checkpoint.

    Replays all pending transformations from the change log onto the original
    file and creates a checkpoint record marking the save point.
    """
    project = get_project_or_404(project_id, db)

    # Load original file for replaying transformations
    original_path = get_original_path(project.file_path)
    df = read_csv_safe(original_path)

    # Get all unapplied logs for this project
    logs = (
        db.query(models.ProjectChangeLog)
        .filter(
            models.ProjectChangeLog.project_id == project_id,
            models.ProjectChangeLog.applied == False,  # noqa: E712
        )
        .order_by(models.ProjectChangeLog.timestamp)
        .all()
    )

    # Replay each logged transformation on the original
    for log in logs:
        df = apply_logged_transformation(df, log.action_type, log.action_details)

    # Write transformations to the working copy (.copy.csv), not the original dataset.
    # original_path must remain immutable as the baseline used for transformation replay.
    assert project.file_path != str(original_path), (
        "Invariant violation: attempted to write transformed data to original_path."
    )
    save_csv_safe(df, project.file_path)

    # Create checkpoint (marks logs as applied)
    checkpoint = create_checkpoint(db, project_id, commit_message)

    resp = dataframe_to_response(df)
    logger.info("Project saved: id=%s, checkpoint=%s", project_id, checkpoint.id)
    return {
        "filename": project.name,
        "file_path": str(project.file_path),
        "project_id": project.project_id,
        **resp,
    }


@router.post("/{project_id}/revert", response_model=schemas.ProjectResponse)
async def revert_to_checkpoint(
    project_id: uuid.UUID,
    checkpoint_id: uuid.UUID = None,
    db: Session = Depends(database.get_db),
):
    """Revert project to its original state or to a specific checkpoint.

    When checkpoint_id is provided, replays only the logs up to and including
    that checkpoint onto the original file. When None, reverts to the original
    uploaded state.
    """
    project = get_project_or_404(project_id, db)

    original_path = get_original_path(project.file_path)
    df = read_csv_safe(original_path)

    if checkpoint_id is not None:
        checkpoint = (
            db.query(models.Checkpoint)
            .filter(
                models.Checkpoint.id == checkpoint_id,
                models.Checkpoint.project_id == project_id,
            )
            .first()
        )
        if not checkpoint:
            raise HTTPException(status_code=404, detail="Checkpoint not found")

        # Find all checkpoint IDs created at or before the target checkpoint
        eligible_checkpoint_ids = [
            c.id
            for c in db.query(models.Checkpoint)
            .filter(
                models.Checkpoint.project_id == project_id,
                models.Checkpoint.created_at <= checkpoint.created_at,
            )
            .all()
        ]

        logs = (
            db.query(models.ProjectChangeLog)
            .filter(
                models.ProjectChangeLog.project_id == project_id,
                models.ProjectChangeLog.checkpoint_id.in_(eligible_checkpoint_ids),
                models.ProjectChangeLog.applied == True,  # noqa: E712
            )
            .order_by(models.ProjectChangeLog.timestamp)
            .all()
        )

        for log in logs:
            df = apply_logged_transformation(df, log.action_type, log.action_details)

    save_csv_safe(df, project.file_path)
    db.commit()

    resp = dataframe_to_response(df)
    logger.info("Project reverted: id=%s, checkpoint_id=%s", project_id, checkpoint_id)
    return {
        "filename": project.name,
        "file_path": project.file_path,
        "project_id": project.project_id,
        **resp,
    }


def _download_name(project_name: str, extension: str) -> str:
    safe_name = re.sub(r"[^\w.\-]+", "_", project_name).strip("._") or "project"
    return f"{safe_name}.{extension}"


def _content_disposition(filename: str) -> dict[str, str]:
    return {"Content-Disposition": f'attachment; filename="{filename}"'}


@router.get("/{project_id}/export")
async def export_project(
    project_id: uuid.UUID,
    format: schemas.ExportFormat = schemas.ExportFormat.csv,
    db: Session = Depends(database.get_db),
):
    """Download the current working copy of a project in the requested format."""
    project = get_project_or_404(project_id, db)
    if format == schemas.ExportFormat.csv:
        return FileResponse(
            project.file_path,
            media_type=EXPORT_MEDIA_TYPES[format],
            filename=_download_name(project.name, "csv"),
        )

    df = read_csv_safe(project.file_path)
    payload = dataframe_to_bytes(df, format.value)
    return Response(
        content=payload,
        media_type=EXPORT_MEDIA_TYPES[format],
        headers=_content_disposition(_download_name(project.name, format.value)),
    )


@router.get("/{project_id}/quality-report")
async def project_quality_report(
    project_id: uuid.UUID,
    format: schemas.ReportFormat = schemas.ReportFormat.html,
    db: Session = Depends(database.get_db),
):
    """Generate a quality report for the current working copy."""
    project = get_project_or_404(project_id, db)
    df = read_csv_safe(project.file_path)
    payload = build_quality_report(project.name, df, format)
    return Response(
        content=payload,
        media_type=REPORT_MEDIA_TYPES[format],
        headers=_content_disposition(_download_name(f"{project.name}_quality_report", format.value)),
    )


@router.delete("/{project_id}")
async def delete_project_endpoint(project_id: uuid.UUID, db: Session = Depends(database.get_db)):
    """Delete a project and its associated files."""
    project = get_project_or_404(project_id, db)
    delete_project_files(project.file_path)
    delete_project(db, project)
    return {"success": True, "message": "Project deleted"}
